### DISCUSSION TITLE
CORS policy: Cross origin requests

### DISCUSSION THREAD

**Discussion by aicameron10** [KevinnZou/compose-webview-multiplatform#38] [Q&A] [answered]
Access to XMLHttpRequest at 'file:///android_asset/book.epub' from origin
has been blocked by CORS policy: Cross origin requests
Android 33

How can this be solved ?

code snippets: 
val webViewState = rememberWebViewStateWithHTMLFile(
        fileName = "index.html"
    )

 webViewState.webSettings.apply {
        isJavaScriptEnabled = true
        logSeverity = KLogSeverity.Debug
        androidWebSettings.apply {
            isAlgorithmicDarkeningAllowed = true
            safeBrowsingEnabled = true
            allowFileAccess = true
        }
        iOSWebSettings.apply { }
    }

**Comment by KevinnZou** [comment, marked as answer]
Hi! According to this [post](https://stackoverflow.com/questions/37984896/android-webview-no-access-control-allow-origin-header-is-present-on-the-reque), we may need to set `allowFileAccessFromFileURLs` and `allowUniversalAccessFromFileURLs` to true. However, it is not currently open for customization. I apologize for any inconvenience caused by this issue.

**Reply by aicameron10** [reply]
Yes that is correct, or use , androidx.webkit.WebViewAssetLoader, but then need to have WebViewClient() option to intercept ,
is there any plans or time line to these options, allowFileAccessFromFileURLs and allowUniversalAccessFromFileURLs ?

**Reply by KevinnZou** [reply]
Since WebViewClient is an Android-only concept, I am not planning to open it for intercept. For the rest two, I think they will be released next week.

By the way, did you encounter the issue on the iOS and Desktop side?

**Reply by aicameron10** [reply]
oh wow thank you for the response, next week already , that is great, and great work on this so far, really is helpful , i will test iOS soon and can let you know.

**Reply by KevinnZou** [reply]
@aicameron10 Hi, I have pushed the changes to this [branch](https://github.com/KevinnZou/compose-webview-multiplatform/tree/feature/allow_file_access). I will merge it after you confirm the behavior on iOS.

**Reply by aicameron10** [reply]
wow that was quick, i had some issues with the ios Simulator and file system to test this in a web-view, so i have got a device, just waiting for the  usb-c to lightning cable i ordered so that i can test on a physical device , as it is basically in the JS, loading files from the devices internal storage, in Android that is where i got the problem with file://. i will try update back here as soon as i can connect the physical device.

**Reply by aicameron10** [reply]
Hi
I do see the html and JS loading 

🔵 (ComposeWebView) Observe URL Changed file:///private/var/containers/Bundle/Application/9269C984-9614-05E71733856C/My%20application.app/compose-resources/assets/index.html
but the file that loads in the Js function 
file:///private/var/mobile/Containers/Data/Application/2EECAF29-91BD14529B/tmp/Lavinia.epub

does not seem to do anything , i cant really see in the logs, if its a setting , javascript is enabled.

does iOS have any web-settings that can be added like android ?
```
 iOSWebSettings.apply {
        }

```
i think maybe this is needed like the native web-view
```
 let webConfiguration = WKWebViewConfiguration()
        
        // Enable file access from file URLs
        webConfiguration.preferences.javaScriptEnabled = true
        webConfiguration.setValue(true, forKey: "allowFileAccessFromFileURLs")

        webView = WKWebView(frame: .zero, configuration: webConfiguration)
        view.addSubview(webView)
```

how i set state:
 ```
webViewState.webSettings.apply {
        isJavaScriptEnabled = true
        logSeverity = KLogSeverity.Debug
        androidWebSettings.apply {
            isAlgorithmicDarkeningAllowed = true
            supportZoom = true
            safeBrowsingEnabled = true
            allowFileAccess = true

        }
        iOSWebSettings.apply {
        }
```

**Reply by KevinnZou** [reply]
Thank you for your help! I just discovered that we need to configure WKWebViewConfiguration in order to grant it access to the file. 

I have fixed it and pushed to that [branch](https://github.com/KevinnZou/compose-webview-multiplatform/tree/feature/allow_file_access). Could you please have a test on it? You just need to set `webSettings` like that: 
```
webViewState.webSettings.apply {
        isJavaScriptEnabled = true
        logSeverity = KLogSeverity.Debug
        allowFileAccessFromFileURLs = true
        allowUniversalAccessFromFileURLs = true
        androidWebSettings.apply {
            isAlgorithmicDarkeningAllowed = true
            safeBrowsingEnabled = true
            allowFileAccess = true
        }
    }
```
Thanks for your help again!

**Reply by aicameron10** [reply]
Hi

Thanks again for the quick response and action to this.

Android and iOS are working with this change

**Reply by andrew-debijenkorf** [reply]
Hi 
Will this be added to 1.8.0 ?

**Reply by KevinnZou** [reply]
It should be added to 1.7.2 and will be released next week.

**Reply by andrew-debijenkorf** [reply]
Thank you
